File: green/src/clj/io/github/getcolors/umami/compute.clj
(ns io.github.getcolors.umami.compute
  (:require [clojure.string :as str] [clojure.walk :as walk] [cheshire.core :as json] [babashka.fs :as fs]
            [io.github.getcolors.compute :as compute]
            [io.github.getcolors.compute-deployment-request :as request]
            [io.github.getcolors.compute-planning :as planning]
            [io.github.getcolors.compute-orchestration :as orchestration]
            [io.github.getcolors.compute-inspection :as inspection]))
(def topology [{:role nil :count 1}])
(defn requirements [opts]
  (let [ssh (request/source-cidrs opts "ssh-sources" "umami-ssh-sources") http (request/source-cidrs opts "http-sources" "umami-http-sources")]
    {:single_host true :legacy_state_keys [(str (:profile opts) "/umami-infrastructure.tfstate")]
     :security {:ingress (into [{:id "ssh" :protocol "tcp" :from_port 22 :to_port 22 :sources ssh}]
                              (when (seq http) (for [port [80 443]] {:id (str "http-" port) :protocol "tcp" :from_port port :to_port port :sources http}))) :egress "all" :private_filter false}}))
(defn errors [opts] (try (planning/validate-deployment opts topology (requirements opts)) (compute/backend-plan opts (str (:profile opts) "/compute/shared.tfstate")) [] (catch Exception _ ["invalid compute deployment requirements"])))
(defn planned [opts] (planning/plan-deployment opts topology (requirements opts)))
(defn node [result] (let [nodes (get-in result [:cluster :nodes])] (when-not (and (= 1 (count nodes)) (:ip (first nodes))) (throw (ex-info "compute node unavailable" {}))) (first nodes)))
(defn attach [opts result]
  (cond
    (not (contains? #{"planned" "ready" "present" "destroyed"} (:status result))) (assoc opts :green/exit 1 :green/err (if (seq (:errors result)) (str/join "\n" (:errors result)) "compute lifecycle refused"))
    (= "destroyed" (:status result)) (assoc opts :green/exit 0 :umami/already-destroyed true)
    :else (let [path (get-in result [:key :private_key_path]) path (if (and path (or (= :build (:green/event opts)) (:green/dry-run opts))) (str/replace path "$HOME" "/home/build-placeholder") path)]
            (cond-> (merge opts (node result) {:colors-compute/cluster (:cluster result) :green/exit 0}) path (assoc :ssh-private-key-path path)))))
(defn- compute-json [value indent]
  (let [padding #(apply str (repeat % " "))]
    (cond
      (map? value) (if (empty? value) "{}"
                      (str "{\n" (str/join ",\n" (for [[key item] (sort-by (comp name key) value)]
                                                       (str (padding (+ indent 2)) (json/generate-string key) ": " (compute-json item (+ indent 2)))))
                           "\n" (padding indent) "}"))
      (sequential? value) (if (empty? value) "[]"
                              (str "[\n" (str/join ",\n" (map #(str (padding (+ indent 2)) (compute-json % (+ indent 2))) value)) "\n" (padding indent) "]"))
      :else (json/generate-string value))))

(defn infrastructure-step [opts]
  (try (let [planning? (or (= :build (:green/event opts)) (:green/dry-run opts)) result (if planning? (planned opts) (orchestration/orchestrate opts topology (requirements opts)))]
         (when planning? (let [root (fs/path (:workdir opts) (:profile opts) "compute") stages (assoc (into {} (for [[id docs] (get-in result [:documents :nodes])] [(str "nodes/" (name id)) docs])) "shared" (get-in result [:documents :shared]))]
                           (doseq [[stage docs] stages
                                   :let [state-key (if (= stage "shared") (get-in result [:state_keys :shared]) (get-in result [:state_keys :nodes (subs stage 6)]))
                                         docs (assoc docs "backend.tf.json" (:config (compute/backend-plan opts state-key))) ]
                                   [filename document] docs]
                             (fs/create-dirs (fs/path root stage))
                             (spit (str (fs/path root stage (name filename))) (str (compute-json document 0) "\n")))))
         (attach opts result)) (catch Exception _ (assoc opts :green/exit 1 :green/err "invalid compute deployment requirements"))))
(defn load-step [opts env] (attach opts (inspection/read-deployment opts env nil (requirements opts))))

File: blue/src/package_umami_blue/compute.py
"""One node through the shared compute lifecycle; application policy only."""
import json
from pathlib import Path
from colors_compute import orchestrate, plan_deployment, read_deployment, source_cidrs
from colors_compute.rendering import backend_plan
from colors_compute.planning import validate_deployment

TOPOLOGY = [{'role': None, 'count': 1}]

def requirements(opts):
    ssh = source_cidrs(opts, 'ssh-sources', 'umami-ssh-sources')
    http = source_cidrs(opts, 'http-sources', 'umami-http-sources')
    return {'single_host': True, 'legacy_state_keys': [opts['profile'] + '/umami-infrastructure.tfstate'],
            'security': {'ingress': [{'id': 'ssh', 'protocol': 'tcp', 'from_port': 22, 'to_port': 22, 'sources': ssh}]
                        + [{'id': 'http-' + str(port), 'protocol': 'tcp', 'from_port': port, 'to_port': port, 'sources': http} for port in (80,443) if http], 'egress': 'all', 'private_filter': False}}

def errors(opts):
    try:
        validate_deployment(opts, TOPOLOGY, requirements(opts))
        backend_plan(opts, opts['profile'] + '/compute/shared.tfstate')
        return []
    except Exception:
        return ['invalid compute deployment requirements']

def planned(opts):
    return plan_deployment(opts, TOPOLOGY, requirements(opts))

def node(result):
    nodes = result.get('cluster', {}).get('nodes', [])
    if len(nodes) != 1 or not nodes[0].get('ip'):
        raise ValueError('compute node unavailable')
    return nodes[0]

def attach(opts, result):
    if result.get('status') not in ('planned', 'ready', 'present', 'destroyed'):
        return {**opts, 'blue/exit': 1, 'blue/err': '\n'.join(result.get('errors', [])) or 'compute lifecycle refused'}
    if result['status'] == 'destroyed':
        return {**opts, 'blue/exit': 0, 'umami/already-destroyed': True}
    params = node(result)
    path = result.get('key', {}).get('private_key_path')
    if path and (opts.get('blue/event') == 'build' or opts.get('blue/dry-run')):
        path = path.replace('$HOME', '/home/build-placeholder')
    return {**opts, **params, 'colors-compute/cluster': result['cluster'], 'blue/exit': 0,
            **({'ssh-private-key-path': path} if path else {})}

async def infrastructure_step(opts):
    try:
        planning = opts.get('blue/event') == 'build' or opts.get('blue/dry-run')
        result = planned(opts) if planning else await orchestrate(opts, TOPOLOGY, requirements(opts))
        if planning:
            root = Path(opts['workdir']) / opts['profile'] / 'compute'
            stages = {'shared': result['documents']['shared'], **{'nodes/'+key: docs for key, docs in result['documents']['nodes'].items()}}
            for stage, documents in stages.items():
                directory = root / stage
                directory.mkdir(parents=True, exist_ok=True)
                state_key = result["state_keys"]["shared"] if stage == "shared" else result["state_keys"]["nodes"][stage.removeprefix("nodes/")]
                documents = {**documents, "backend.tf.json": backend_plan(opts, state_key)["config"]}
                for filename, document in documents.items():
                    (directory / filename).write_text(json.dumps(document, sort_keys=True, indent=2) + '\n')
        return attach(opts, result)
    except Exception:
        return {**opts, 'blue/exit': 1, 'blue/err': 'invalid compute deployment requirements'}

async def load(opts, environment=None):
    return attach(opts, await read_deployment(opts, environment, None, requirements(opts)))
